Tests/EndpointClosureSpec.swift

```diff
@@ -0,0 +1,83 @@
+                let task: Task
+
+                switch target.task {
+                case let .uploadMultipart(multipartFormData):
+                    let additional = Moya.MultipartFormData(provider: .data("test2".data(using: .utf8)!), name: "test2")
+                    var newMultipartFormData = multipartFormData
+                    newMultipartFormData.append(additional)
+                    task = .uploadMultipart(newMultipartFormData)
+                default:
+                    task = target.task
+                }
+
+                return Endpoint<HTTPBin>(url: URL(target: target).absoluteString, sampleResponseClosure: {.networkResponse(200, target.sampleData)}, method: target.method, task: task, httpHeaderFields: target.headers)
+            }
+            provider = MoyaProvider<HTTPBin>(endpointClosure: endpointClosure, manager: sessionManager)
+        }
+
+        it("returns a new endpoint for adding(newHTTPHeaderFields:)") {
+            let multipartData1 = Moya.MultipartFormData(provider: .data("test1".data(using: .utf8)!), name: "test1")
+            let multipartData2 = Moya.MultipartFormData(provider: .data("test2".data(using: .utf8)!), name: "test2")
+
+            let providedMultipartData = [multipartData1]
+            let sentMultipartData = [multipartData1, multipartData2]
+
+            _ = provider.request(.uploadMultipart(providedMultipartData, nil)) { _ in }
+            let stringData1 = String(data: try! sessionManager.uploadMultipartFormData!.encode(), encoding: .utf8)
+            let stringData2 = String(data: try! sessionManager.convertMoyaMultipartFormDataToAlamofire(sentMultipartData).encode(), encoding: .utf8)
+
+            let formData1 = "data; name=\"test1\"\r\n\r\ntest1\r\n"
+            let formData2 = "data; name=\"test2\"\r\n\r\ntest2\r\n"
+
+            let splitString1 = stringData1?.split(separator: "-").map { String($0) }
+            let splitString2 = stringData2?.split(separator: "-").map { String($0) }
+
+            expect(splitString1).to(contain(formData1))
+            expect(splitString1).to(contain(formData2))
+            expect(splitString2).to(contain(formData1))
+            expect(splitString2).to(contain(formData2))
+
+        }
+    }
+}
+
+final class SessionManagerMock: SessionManager {
+
+    var uploadMultipartFormData: Alamofire.MultipartFormData?
+
+    override func upload(multipartFormData: @escaping (Alamofire.MultipartFormData) -> Void, usingThreshold encodingMemoryThreshold: UInt64, with urlRequest: URLRequestConvertible, encodingCompletion: ((SessionManager.MultipartFormDataEncodingResult) -> Void)?) {
+        let uploadMultipartFormData = Alamofire.MultipartFormData()
+        multipartFormData(uploadMultipartFormData)
+        self.uploadMultipartFormData = uploadMultipartFormData
+    }
+
+    func convertMoyaMultipartFormDataToAlamofire(_ multipartBody: [Moya.MultipartFormData]) -> Alamofire.MultipartFormData {
```

Comment:
I've added this method here, but maybe we can add it as an internal extension to `Moya.MultipartFormData`?

Reply:
I think it is better to keep this in the test target if we're not using it in the core library?

Reply:
[Actually, we use it in the core as well!](https://github.com/Moya/Moya/blob/f071d9703fbbd3f4a309fbbe32eae3ac6ca2f32d/Sources/Moya/MoyaProvider%2BInternal.swift#L154). If you have time to refactor this a little bit, go for it, if not, let's just make this an extension to `MultipartFormData` and add an issue for a little refactor?

Reply:
Ok, thanks for pointing that out. I'll take care of it tonight 👍